File: apps/composeApp/src/commonMain/kotlin/com/arikachmad/pebblerun/ui/workout/WorkoutViewModel.kt
package com.arikachmad.pebblerun.ui.workout

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arikachmad.pebblerun.domain.model.WorkoutSession
import com.arikachmad.pebblerun.domain.model.WorkoutStatus
import com.arikachmad.pebblerun.domain.usecase.StartWorkoutUseCase
import com.arikachmad.pebblerun.domain.usecase.StopWorkoutUseCase
import com.arikachmad.pebblerun.domain.usecase.UpdateWorkoutDataUseCase
import com.arikachmad.pebblerun.data.repository.MockWorkoutRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for workout screen implementing MVVM pattern
 * Handles workout session state management and user interactions
 */
class WorkoutViewModel : ViewModel() {

    // Mock dependencies for UI testing
    private val mockRepository = MockWorkoutRepository()
    private val startWorkoutUseCase = StartWorkoutUseCase(mockRepository)
    private val stopWorkoutUseCase = StopWorkoutUseCase(mockRepository)
    private val updateWorkoutDataUseCase = UpdateWorkoutDataUseCase(mockRepository)

    private val _uiState = MutableStateFlow(WorkoutUiState())
    val uiState: StateFlow<WorkoutUiState> = _uiState.asStateFlow()
    
    private var currentSession: WorkoutSession? = null

    fun startWorkout() {
        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(isLoading = true, error = null)
                
                val session = startWorkoutUseCase.execute()
                currentSession = session
                
                _uiState.value = _uiState.value.copy(
                    currentSession = session,
                    isLoading = false,
                    isWorkoutActive = true
                )
                
                startDataSimulation()
                
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to start workout"
                )
            }
        }
    }

    fun pauseWorkout() {
        viewModelScope.launch {
            currentSession?.let { session ->
                try {
                    startWorkoutUseCase.pauseWorkout(session.id)
                    
                    val updatedSession = session.copy(status = WorkoutStatus.PAUSED)
                    currentSession = updatedSession
                    
                    _uiState.value = _uiState.value.copy(
                        currentSession = updatedSession,
                        isWorkoutActive = false
                    )
                    
                } catch (e: Exception) {
                    _uiState.value = _uiState.value.copy(error = e.message)
                }
            }
        }
    }

    fun resumeWorkout() {
        viewModelScope.launch {
            currentSession?.let { session ->
                try {
                    startWorkoutUseCase.resumeWorkout(session.id)
                    
                    val updatedSession = session.copy(status = WorkoutStatus.ACTIVE)
                    currentSession = updatedSession
                    
                    _uiState.value = _uiState.value.copy(
                        currentSession = updatedSession,
                        isWorkoutActive = true
                    )
                    
                    startDataSimulation()
                    
                } catch (e: Exception) {
                    _uiState.value = _uiState.value.copy(error = e.message)
                }
            }
        }
    }

    fun stopWorkout() {
        viewModelScope.launch {
            currentSession?.let { session ->
                try {
                    _uiState.value = _uiState.value.copy(isLoading = true)
                    
                    val completedSession = stopWorkoutUseCase.execute(session.id)
                    
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        currentSession = completedSession,
                        isWorkoutActive = false,
                        showWorkoutComplete = true
                    )
                    
                    currentSession = null
                    
                } catch (e: Exception) {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = e.message
                    )
                }
            }
        }
    }

    fun dismissWorkoutComplete() {
        _uiState.value = _uiState.value.copy(
            showWorkoutComplete = false,
            currentSession = null
        )
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    private fun startDataSimulation() {
        viewModelScope.launch {
            while (currentSession?.status == WorkoutStatus.ACTIVE) {
                currentSession?.let { session ->
                    val updatedSession = session.copy(
                        totalDuration = session.totalDuration + 1000, // +1 second
                        totalDistance = session.totalDistance + 2.5, // +2.5 meters
                        averagePace = if (session.totalDistance > 0) 
                            (session.totalDuration / 1000.0) / (session.totalDistance / 1000.0) * 1000 
                        else 0.0,
                        averageHeartRate = session.hrSamples.lastOrNull()?.bpm ?: 140,
                        calories = (session.totalDuration / 1000 * 0.2).toInt()
                    )
                    
                    currentSession = updatedSession
                    _uiState.value = _uiState.value.copy(currentSession = updatedSession)
                }
                
                kotlinx.coroutines.delay(1000)
            }
        }
    }
}

data class WorkoutUiState(
    val isLoading: Boolean = false,
    val currentSession: WorkoutSession? = null,
    val isWorkoutActive: Boolean = false,
    val showWorkoutComplete: Boolean = false,
    val error: String? = null
)

File: apps/iosApp/iosApp/ViewModel/WorkoutViewModel.swift
import Foundation
import Combine
import shared

/**
 * iOS-specific ViewModel for WorkoutView using ObservableObject pattern.
 * Implements TASK-019: Create iOS-specific ObservableObject classes for KMP integration.
 * Follows PAT-002: ObservableObject pattern for iOS SwiftUI integration.
 */
@MainActor
class WorkoutViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var isWorkoutActive = false
    @Published var workoutDuration = "00:00:00"
    @Published var distance = "0.00 km"
    @Published var pace = "--:--"
    @Published var currentHeartRate = "--"
    @Published var averageHeartRate = "--"
    @Published var maxHeartRate = "--"
    @Published var isPebbleConnected = false
    @Published var showError = false
    @Published var errorMessage = ""
    
    // MARK: - Computed Properties
    var workoutStatusText: String {
        if isWorkoutActive {
            return "Workout Active"
        } else {
            return "Ready to Start"
        }
    }
    
    var workoutStartTime: String {
        // Format the start time - this would come from the actual workout session
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return formatter.string(from: Date())
    }
    
    var pebbleStatusText: String {
        return isPebbleConnected ? "Pebble Connected" : "Pebble Disconnected"
    }
    
    // MARK: - Private Properties
    private var workoutTrackingService: WorkoutTrackingService
    private var cancellables = Set<AnyCancellable>()
    private var workoutTimer: Timer?
    private var startTime: Date?
    
    // KMP Integration - Use cases from shared domain
    private let startWorkoutUseCase: StartWorkoutUseCase
    private let stopWorkoutUseCase: StopWorkoutUseCase
    private let updateWorkoutDataUseCase: UpdateWorkoutDataUseCase
    
    // MARK: - Initialization
    init() {
        // Use dependency injection from IOSContainer
        let container = IOSContainer.shared
        
        self.startWorkoutUseCase = container.startWorkoutUseCase
        self.stopWorkoutUseCase = container.stopWorkoutUseCase
        self.updateWorkoutDataUseCase = container.updateWorkoutDataUseCase
        
        // Initialize iOS-specific service
        self.workoutTrackingService = container.workoutTrackingService
        
        setupBindings()
    }
    
    // MARK: - Setup
    private func setupBindings() {
        // Observe workout tracking service state
        workoutTrackingService.$serviceState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.updateUIForServiceState(state)
            }
            .store(in: &cancellables)
        
        // Observe current workout session
        workoutTrackingService.$currentSession
            .receive(on: DispatchQueue.main)
            .sink { [weak self] session in
                self?.updateUIForSession(session)
            }
            .store(in: &cancellables)
        
        // Check Pebble connection status periodically
        Timer.publish(every: 5.0, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                self?.checkPebbleConnection()
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Public Methods
    func onViewAppear() {
        checkPebbleConnection()
        loadCurrentWorkoutState()
    }
    
    func startWorkout() {
        Task {
            do {
                isWorkoutActive = true
                startTime = Date()
                startWorkoutTimer()
                
                // Start tracking service
                try await workoutTrackingService.startWorkout()
                
            } catch {
                handleError(error)
                isWorkoutActive = false
            }
        }
    }
    
    func stopWorkout() {
        Task {
            do {
                // Stop tracking service
                try await workoutTrackingService.stopWorkout()
                
                isWorkoutActive = false
                stopWorkoutTimer()
                resetWorkoutDisplay()
                
            } catch {
                handleError(error)
            }
        }
    }
    
    func pauseWorkout() {
        Task {
            do {
                // Pause tracking service
                try await workoutTrackingService.pauseWorkout()
                stopWorkoutTimer()
                
            } catch {
                handleError(error)
            }
        }
    }
    
    func connectToPebble() {
        Task {
            do {
                try await workoutTrackingService.connectToPebble()
                isPebbleConnected = true
            } catch {
                handleError(error)
            }
        }
    }
    
    func dismissError() {
        showError = false
        errorMessage = ""
    }
    
    // MARK: - Private Methods
    private func updateUIForServiceState(_ state: WorkoutTrackingService.ServiceState) {
        switch state {
        case .idle:
            isWorkoutActive = false
        case .tracking:
            isWorkoutActive = true
        case .paused:
            // Keep active but stop timer
            break
        case .error(let message):
            handleError(NSError(domain: "WorkoutService", code: -1, userInfo: [NSLocalizedDescriptionKey: message]))
        }
    }
    
    private func updateUIForSession(_ session: WorkoutSession?) {
        guard let session = session else {
            resetWorkoutDisplay()
            return
        }
        
        // Update distance
        let distanceKm = session.totalDistance / 1000.0
        distance = String(format: "%.2f km", distanceKm)
        
        // Update pace (assuming session.averagePace is in seconds per km)
        if session.averagePace > 0 {
            let minutes = Int(session.averagePace) / 60
            let seconds = Int(session.averagePace) % 60
            pace = String(format: "%d:%02d", minutes, seconds)
        }
        
        // Update heart rate
        if session.averageHeartRate > 0 {
            currentHeartRate = "\(session.averageHeartRate)"
            averageHeartRate = "\(session.averageHeartRate)"
            maxHeartRate = "\(session.maxHeartRate)"
        }
    }
    
    private func startWorkoutTimer() {
        workoutTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.updateWorkoutDuration()
        }
    }
    
    private func stopWorkoutTimer() {
        workoutTimer?.invalidate()
        workoutTimer = nil
    }
    
    private func updateWorkoutDuration() {
        guard let startTime = startTime else { return }
        
        let elapsed = Date().timeIntervalSince(startTime)
        let hours = Int(elapsed) / 3600
        let minutes = (Int(elapsed) % 3600) / 60
        let seconds = Int(elapsed) % 60
        
        workoutDuration = String(format: "%02d:%02d:%02d", hours, minutes, seconds)
    }
    
    private func resetWorkoutDisplay() {
        workoutDuration = "00:00:00"
        distance = "0.00 km"
        pace = "--:--"
        currentHeartRate = "--"
        averageHeartRate = "--"
        maxHeartRate = "--"
        startTime = nil
    }
    
    private func checkPebbleConnection() {
        // This would integrate with the actual Pebble bridge
        // For now, simulate connection status
        isPebbleConnected = workoutTrackingService.isPebbleConnected
    }
    
    private func loadCurrentWorkoutState() {
        // Load any existing workout state from the service
        if let currentSession = workoutTrackingService.currentSession {
            updateUIForSession(currentSession)
            isWorkoutActive = workoutTrackingService.serviceState.isActive
            
            if isWorkoutActive {
                startTime = Date(timeIntervalSince1970: TimeInterval(currentSession.startTime.toEpochMilliseconds() / 1000))
                startWorkoutTimer()
            }
        }
    }
    
    private func handleError(_ error: Error) {
        errorMessage = error.localizedDescription
        showError = true
        print("WorkoutViewModel Error: \(error)")
    }
}

// MARK: - Mock Implementations
// These would be replaced with actual KMP integrations

private class MockWorkoutRepository: WorkoutRepository {
    func createSession(session: WorkoutSession) async -> DomainResult<WorkoutSession> {
        return DomainResult.Success(data: session)
    }
    
    func updateSession(session: WorkoutSession) async -> DomainResult<WorkoutSession> {
        return DomainResult.Success(data: session)
    }
    
    func getSessionById(id: String) async -> DomainResult<WorkoutSession?> {
        return DomainResult.Success(data: nil)
    }
    
    func getAllSessions(limit: Int32?, offset: Int32, status: WorkoutStatus?, startDate: Instant?, endDate: Instant?) async -> DomainResult<[WorkoutSession]> {
        return DomainResult.Success(data: [])
    }
    
    func observeSessions() -> Flow<[WorkoutSession]> {
        // Return empty flow for mock
        return FlowKt.flowOf([])
    }
    
    func observeSession(id: String) -> Flow<WorkoutSession?> {
        return FlowKt.flowOf(nil)
    }
    
    func deleteSession(id: String) async -> DomainResult<KotlinUnit> {
        return DomainResult.Success(data: KotlinUnit())
    }
    
    func getActiveSession() async -> DomainResult<WorkoutSession?> {
        return DomainResult.Success(data: nil)
    }
    
    func observeActiveSession() -> Flow<WorkoutSession?> {
        return FlowKt.flowOf(nil)
    }
    
    func completeSession(id: String, endTime: Instant, finalStats: WorkoutSessionStats) async -> DomainResult<WorkoutSession> {
        // Create a mock completed session
        let session = WorkoutSession(
            id: id,
            startTime: endTime,
            endTime: endTime,
            status: WorkoutStatus.completed,
            totalDuration: finalStats.totalDuration,
            totalDistance: finalStats.totalDistance,
            averagePace: 300.0, // 5:00/km
            averageHeartRate: finalStats.averageHeartRate,
            maxHeartRate: finalStats.maxHeartRate,
            minHeartRate: 60,
            calories: finalStats.calories,
            geoPoints: [],
            hrSamples: [],
            notes: ""
        )
        return DomainResult.Success(data: session)
    }
    
    func getSessionStats(id: String) async -> DomainResult<WorkoutSessionStats?> {
        return DomainResult.Success(data: nil)
    }
    
    func exportSessions(sessionIds: [String]) async -> DomainResult<String> {
        return DomainResult.Success(data: "mock-export-data")
    }
}

private class MockLocationProvider: LocationProvider {
    func startLocationUpdates() async throws {
        // Mock implementation
    }
    
    func stopLocationUpdates() {
        // Mock implementation
    }
    
    func getCurrentLocation() async -> GeoPoint? {
        return nil
    }
    
    func observeLocationUpdates() -> Flow<GeoPoint> {
        return FlowKt.flowOf(GeoPoint(lat: 0.0, lon: 0.0, timestamp: 0))
    }
    
    func hasLocationPermission() -> Bool {
        return true
    }
    
    func requestLocationPermission() async -> Bool {
        return true
    }
}

private class MockPebbleTransport: PebbleTransport {
    func connect() async throws {
        // Mock implementation
    }
    
    func disconnect() {
        // Mock implementation
    }
    
    func sendMessage(message: [String: Any]) async throws {
        // Mock implementation
    }
    
    func observeMessages() -> Flow<[String: Any]> {
        return FlowKt.flowOf([:])
    }
    
    func isConnected() -> Bool {
        return false
    }
}
